React 19 support

Followup to https://github.com/remarkjs/react-markdown/issues/871

`react-markdown` relies on `JSX` being in the global namespace, which has been removed in `@types/react^19`. 

Per @ChristianMurphy [here](https://github.com/remarkjs/react-markdown/issues/871#issuecomment-2551504752):

> A new issue would be better, we won't support React 19 RC, only GA.

[React 19 is now GA](https://react.dev/blog/2024/12/05/react-19), so this issue is to track.

@wooorm If I upgrade to React 19 in this repo and `npm run build`, I get the following error:

```
> react-markdown@9.0.1 build
> tsc --build --clean && tsc --build && type-coverage

node_modules/hast-util-to-jsx-runtime/lib/index.d.ts:12:62 - error TS2503: Cannot find namespace 'JSX'.

12 export function toJsxRuntime(tree: Nodes, options: Options): JSX.Element;
```

In https://github.com/syntax-tree/hast-util-to-jsx-runtime/commit/ae638ba52762906fe094d74f82ba835ce23a2621 you provide the solution of declaring a global JSX namespace. Would you recommend doing that here?

--- a/index.js
+++ b/index.js
@@ -1,7 +1,7 @@
 /**
- * @typedef {import('hast-util-to-jsx-runtime').ExtraProps} ExtraProps
  * @typedef {import('./lib/index.js').AllowElement} AllowElement
  * @typedef {import('./lib/index.js').Components} Components
+ * @typedef {import('./lib/index.js').ExtraProps} ExtraProps
  * @typedef {import('./lib/index.js').Options} Options
  * @typedef {import('./lib/index.js').UrlTransform} UrlTransform
  */

--- a/lib/index.js
+++ b/lib/index.js
@@ -1,7 +1,6 @@
 /**
  * @import {Element, ElementContent, Nodes, Parents, Root} from 'hast'
- * @import {Components as JsxRuntimeComponents} from 'hast-util-to-jsx-runtime'
- * @import {ReactElement} from 'react'
+ * @import {ComponentProps, ElementType, ReactElement} from 'react'
  * @import {Options as RemarkRehypeOptions} from 'remark-rehype'
  * @import {BuildVisitor} from 'unist-util-visit'
  * @import {PluggableList} from 'unified'
@@ -21,7 +20,16 @@
  */
 
 /**
- * @typedef {Partial<JsxRuntimeComponents>} Components
+ * @typedef ExtraProps
+ *   Extra fields we pass.
+ * @property {Element | undefined} [node]
+ *   passed when `passNode` is on.
+ */
+
+/**
+ * @typedef {{
+ *   [Key in Extract<ElementType, string>]?: ElementType<ComponentProps<Key> & ExtraProps>
+ * }} Components
  *   Map tag names to components.
  */
 
@@ -225,6 +233,9 @@ export function Markdown(options) {
 
   return toJsxRuntime(hastTree, {
     Fragment,
+    // @ts-expect-error
+    // React components are allowed to return numbers,
+    // but not according to the types in hast-util-to-jsx-runtime
     components,
     ignoreInvalidStyle: true,
     jsx,

